flag.go

```diff
@@ -0,0 +1,212 @@
+	TLAStrFileVar(fs, c.TLAVars, "tla-str-file", "Add top-level arg `var=file` string from a file")
+	TLACodeFileVar(fs, c.TLAVars, "tla-code-file", "Add top-level arg `var=file` code from a file")
+
+	// Add short flags. TODO(camh): consider making these optional.
+	StringSliceVar(fs, &c.ImportPath, "J", "Add a library search `dir`")
+	ExtStrVar(fs, c.ExtVars, "V", "Add extVar `var[=str]` (from environment if <str> is omitted)")
+	TLAStrVar(fs, c.TLAVars, "A", "Add top-level arg `var[=str]` (from environment if <str> is omitted)")
+}
+
+// StringSliceVar defines a flag in the given FlagSet with the given name and
+// usage string. The argument p is a pointer to a []string variable in which to
+// store the value of the flag. The value given to each instance of the flag is
+// appended to the slice.
+func StringSliceVar(fs *flag.FlagSet, p *[]string, name, usage string) {
+	fs.Var((*stringSliceValue)(p), name, usage)
+}
+
+// StringSlice defines a flag in the given FlagSet with the given name and
+// usage string. The return value is the address of a []string variable that
+// stores the value of the flag. The value given to each instance of the flag
+// is appended to the slice.
+func StringSlice(fs *flag.FlagSet, name, usage string) *[]string {
+	p := &[]string{}
+	StringSliceVar(fs, p, name, usage)
+	return p
+}
+
+type stringSliceValue []string
+
+func (s *stringSliceValue) Set(v string) error {
+	*s = append(*s, v)
+	return nil
+}
+
+func (s *stringSliceValue) Get() interface{} { return []string(*s) }
+func (s *stringSliceValue) String() string   { return fmt.Sprint(*s) }
+
+// ExtStrVar defines flag with the given name and usage string in the given
+// FlagSet to set a VMVar in the given VMVarMap. The VMVar sets an extVar
+// string literal in a jsonnet VM.
+//
+// The flag value on the command line is parsed as "key[=string]". If "=string"
+// is omitted then key is looked up in the environment for the value. It is an
+// error if "=string" is not provided and the environment variable does not
+// exist.
+//
+// The flag can be repeated multiple times on the command line. Each instance
+// adds a new value to the given VMVarMap.
+func ExtStrVar(fs *flag.FlagSet, m VMVarMap, name, usage string) {
+	fs.Var(vmVarMapValue{m, NewExtStr, true}, name, usage)
+}
+
+// ExtCodeVar defines flag with the given name and usage string in the given
+// FlagSet to set a VMVar in the given VMVarMap. The VMVar sets an extVar code
+// literal in a jsonnet VM.
+//
+// The flag value on the command line is parsed as "key[=code]". If "=code" is
+// omitted then key is looked up in the environment for the value. It is an
+// error if "=code" is not provided and the environment variable does not
+// exist.
+//
+// The flag can be repeated multiple times on the command line. Each instance
+// adds a new value to the given VMVarMap.
+func ExtCodeVar(fs *flag.FlagSet, m VMVarMap, name, usage string) {
+	fs.Var(vmVarMapValue{m, NewExtCode, true}, name, usage)
```

Comment:
fwiw, I was a bit confused reading `vmVarMapValue{m, NewExtCode, true}` - especially the `true`.
By now I've seen a few times the env/no-env | file/no-file configuration so I kinda guessed it, but I wonder if a constructor would help here ala `newVMVarVal` and `newVMFileVal` or so....

Reply:
I think I have two better ideas - thank you for making me think about this more.

1) (this one doesn't work now I think of it) The second field - the `NewExtCode` and co, encodes enough information somehow to determine whether the type of variable supports values coming from the environment. Unfortunately, I'm not sure how to attach that information. I could add a method to the type that New* creates which returns true/false if it can come from the environment. But I need a value of that type before I can call that method, and I don't have a value of that type until I have the string value. chicken and egg. I could create a `type FromEnvFunc func(string) VMVar` and `type NotFromEnvFunc func(string) VMVar` and put methods on those two types that say whether they accept values from the env, but this is getting a bit weird.

2) Why do I even have this flag at all? Why cant the "file" vars take the filename from an environment variable. The standard jsonnet command does not allow it, and I remember thinking about it and thinking that made sense. But right now, it makes no sense. It should be perfectly valid to say `FOO=bar.jsonnet jsonnext --tla-code-file FOO` and it takes the filename for FOOs contents from the environment variable FOO. This will simplify a lot of stuff around this area - not needing to make this distinction seems like a good simplification.

What do you think of idea 2?

Reply:
Ok. I've done idea 2 - I think the code is simpler now and more powerful too. It was worth breaking from the jsonnet implementation here and allowing filenames to come from the environment.

Reply:
yes! why not.